pkg/process/util/api/weighted_queue.go

```diff
@@ -0,0 +1,208 @@
+
+package api
+
+import (
+	"container/list"
+	"sync"
+)
+
+// WeightedItem is an item that has a type and weight that can be added to a WeightedQueue
+type WeightedItem interface {
+	// Weight returns the weight of this item
+	Weight() int64
+	// Type returns the type of this item
+	Type() string
+}
+
+// WeightedQueue is a queue of WeightedItems.
+//
+// The queue is configured with a maximum size (the maximum number of elements allowed in the queue) as well as
+// a maximum weight.  If adding an item to the queue would violate either the max weight or max size, earlier items
+// are purged from the queue until there is room for the newest item.
+//
+// Items added to the queue have a weight and type.  When purging existing items to make room for new, items of the same
+// type being added will be removed first before moving on to other types.
+type WeightedQueue struct {
+	// dataAvailable is a channel used for communication between callers invoking the Add method and callers
+	// blocked on Poll.  If a caller calls Poll and the queue is empty, the call will block waiting for a send
+	// on dataAvailable.  When Add is invoked, it will perform a non-blocking send on dataAvailable to notify the caller
+	// blocked on Poll
+	dataAvailable chan struct{}
+	queue         *list.List
+	mu            sync.Mutex
+
+	maxSize       int
+	maxWeight     int64
+	currentWeight int64
+}
+
+// NewWeightedQueue returns a new WeightedQueue with the given maximum size & weight
+func NewWeightedQueue(maxSize int, maxWeight int64) *WeightedQueue {
+	return &WeightedQueue{
+		dataAvailable: make(chan struct{}),
+		queue:         list.New(),
+		maxSize:       maxSize,
+		maxWeight:     maxWeight,
+		currentWeight: 0,
+	}
+}
+
+// Len returns the number of items in the queue
+func (q *WeightedQueue) Len() int {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.queue.Len()
+}
+
+// Weight returns the current weight of the queue
+func (q *WeightedQueue) Weight() int64 {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.currentWeight
+}
+
+// Poll retrieves the head of the queue or blocks until an item is available.  The provided exit channel can be closed
+// to interrupt the blocking operation.  Returns the head of the queue and true or nil, false if the poll was
+// interrupted by the closing of the exit channel
+func (q *WeightedQueue) Poll(exit chan struct{}) (WeightedItem, bool) {
+	for {
+		select {
+		case <-exit:
+			return nil, false
+		default:
+
+		}
+
+		q.mu.Lock()
+
+		// If the queue is empty, wait for a signal on the dataAvailable channel
+		if q.queue.Len() == 0 {
+			q.mu.Unlock()
+
+			select {
+			case <-q.dataAvailable:
+				continue
+			case <-exit:
+				return nil, false
+			}
+		}
+
+		e := q.queue.Front()
+		item := e.Value.(WeightedItem)
+		q.queue.Remove(e)
+		q.currentWeight -= item.Weight()
+
+		q.mu.Unlock()
+
+		return item, true
+	}
+}
+
+// Add adds the item to the queue.
+func (q *WeightedQueue) Add(item WeightedItem) {
```

Comment:
Would it be beneficial to add return values to this function to indicate the various scenarios that happened(single item too big, item removed, etc) here? I'm think we might log them to know what's happening.

Reply:
Do you have thoughts on what would be good to return?  I don't have a good idea apart from just logging in the queue itself.

Reply:
How about

```suggestion
func (q *WeightedQueue) Add(item WeightedItem) (ok bool, removed int){
```
`ok` indicates whether the item is added at all, `removed` indicates how many are dequeued.

Reply:
We would lose some precision in logging (whether items were removed for size or weight, which type of items etc).  

Do you think thats important to capture?